src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/ProjectMembershipComponent.scala

```diff
@@ -0,0 +1,56 @@
+package org.broadinstitute.dsde.rawls.dataaccess.slick
+
+case class ProjectMembershipRecord(userSubjectId: String, projectName: String)
+
+trait ProjectMembershipComponent {
+  this: DriverComponent
+    with RawlsUserComponent
+    with RawlsBillingProjectComponent =>
+
+  import driver.api._
+
+  class ProjectMembershipTable(tag: Tag) extends Table[ProjectMembershipRecord](tag, "PROJECT_MEMBERSHIP") {
+    def userSubjectId = column[String]("userSubjectId")
+    def projectName = column[String]("projectName")
+
+    def * = (userSubjectId, projectName) <> (ProjectMembershipRecord.tupled, ProjectMembershipRecord.unapply)
+
+    def user = foreignKey("FK_PROJECT_MEMBERSHIP_USER", userSubjectId, rawlsUserQuery)(_.userSubjectId)
+    def project = foreignKey("FK_PROJECT_MEMBERSHIP_PROJECT", projectName, rawlsBillingProjectQuery)(_.projectName)
+    def uniqueIdx = index("IDX_PROJECT_MEMBERSHIP_UNIQUE", (userSubjectId, projectName), unique = true)
+  }
+
+  val projectMembershipQuery = TableQuery[ProjectMembershipTable]
+
+  def projectMembershipBySubjectIdQuery(userSubjectId: String) = Compiled {
+    projectMembershipQuery.filter(_.userSubjectId === userSubjectId)
+  }
+
+  def projectMembershipByProjectNameQuery(projectName: String) = Compiled {
+    projectMembershipQuery.filter(_.projectName === projectName)
+  }
+
+  def saveProjectMembership(membership: ProjectMembershipRecord) = {
+    (projectMembershipQuery += membership) map { _ => membership }
+  }
+
+  def loadProjectMembershipBySubjectId(userSubjectId: String) = {
+    projectMembershipBySubjectIdQuery(userSubjectId).result
+  }
+
+  def loadProjectMembershipByProjectName(projectName: String) = {
+    projectMembershipByProjectNameQuery(projectName).result
+  }
+
+  def deleteProjectMembership(membership: ProjectMembershipRecord) = {
+    projectMembershipQuery.filter(q => q.userSubjectId === membership.userSubjectId && q.projectName === membership.projectName).delete
+  }
+
+  def deleteProjectMembershipBySubjectId(userSubjectId: String) = {
+    projectMembershipBySubjectIdQuery(userSubjectId).delete
+  }
+
+  def deleteProjectMembershipByProjectName(projectName: String) = {
+    projectMembershipByProjectNameQuery(projectName).delete
+  }
```

Comment:
Hm, ok.  This makes me realize that I don't have a clear picture about how we're doing the DAO part of this and I think having a conversation about that would clarify things.  Specifically I would like to understand:

Are DAOs a layer above Data Components?  (Sounds like yes.)

Do we want Components to be invisible to everything except the DAOs?

Since we are implementing Component tables before the DAOs, we don't necessarily know what they will call.  What is a sufficient test of the table without that knowledge?